### PR TITLE
Add Synthetic Data Generation Module

### PR DESCRIPTION
**Pull request by ryantwolf**
## Description

Adds a suite of tools for interacting with LLM services. These LLM services are then used to build synthetic data generation tools and example pipelines following the [Nemotron 340B Technical Report](https://arxiv.org/pdf/2406.11704v1). The prompt templates used in the report are supplied as defaults throughout the code.

### OpenAI API
```python
from nemo_curator import AsyncOpenAIClient
from nemo_curator.synthetic import (
    AsyncNemotronGenerator,
    NemotronGenerator,
)
from openai import OpenAI, AsyncOpenAI

async def demo():
  openai_client = AsyncOpenAI(
      base_url="https://integrate.api.nvidia.com/v1",
      api_key="",
  )
  client = AsyncOpenAIClient(openai_client)
  generator = AsyncNemotronGenerator(client)
  
  model = "nvidia/nemotron-4-340b-instruct"
  model_kwargs = {
      "top_p": 0.7,
      "max_tokens": 1024,
      "seed": 1234,
  }
  
  openlines = await generator.run_open_qa_pipeline(
      n_macro_topics=5,
      n_subtopics=3,
      n_openlines=3,
      n_revisions=2,
      model=model,
      base_model_kwargs=model_kwargs,
      conversion_model_kwargs=model_kwargs,
      ignore_conversion_failure=True,
  )
  
  dialogue = await generator.generate_dialogue(
      openline=openlines[0],
      user_model=model,
      assistant_model=model,
      user_model_kwargs=model_kwargs,
      assistant_model_kwargs=model_kwargs,
  )

  print(dialogue)
```
### NeMo Deploy
```python
from nemo_curator import NemoDeployClient
from nemo_curator.synthetic import (
    AsyncNemotronGenerator,
    NemotronGenerator,
    NemotronFormatter,
)
from nemo.deploy.nlp import NemoQueryLLM

async def demo():
  model = "local_nemotron"
  model_kwargs = {
      "top_p": 0.7,
      "max_tokens": 1024,
      "seed": 1234,
      "conversation_formatter": NemotronFormatter(),
      "stop": ['<extra_id_1>'],
  }
  
  nemo_client = NemoQueryLLM(url="localhost:8000", model_name=model)
  client = NemoDeployClient(nemo_client)
  
  openlines = await generator.run_open_qa_pipeline(
      n_macro_topics=5,
      n_subtopics=3,
      n_openlines=3,
      n_revisions=2,
      model=model,
      base_model_kwargs=model_kwargs,
      conversion_model_kwargs=model_kwargs,
      ignore_conversion_failure=True,
  )
```
## Checklist

- [x] I am familiar with the [Contributing Guide](https://github.com/NVIDIA/NeMo-Curator/blob/main/CONTRIBUTING.md).
- [x] New or Existing tests cover these changes.
- [ ] The documentation is up to date with these changes.

### CODE REVIEW

**Comment by ryantwolf** [comment]
Need to update docs still, but other than that should be good for a first round.

File: nemo_curator/synthetic/prompts.py [hunk 1]
```diff
@@ -0,0 +1,58 @@
+# Copyright (c) 2024, NVIDIA CORPORATION.  All rights reserved.
```
Comment (VibhuJawa) [inline nemo_curator/synthetic/prompts.py:1]: Nit long term we should probably move to using `jinja` for prompt templates, this is fine for now.

File: nemo_curator/synthetic/nemotron.py [hunk 2]
```diff
@@ -0,0 +1,1472 @@
...
+                    continue
+                else:
+                    raise e
+
+        return openlines
+
+
+class NemotronFormatter(ConversationFormatter):
+
+    BASE_PROMPT = "<extra_id_0>System\n\n<extra_id_1>User\n"
```
Comment (VibhuJawa) [inline nemo_curator/synthetic/nemotron.py:1442]: Should this be in prompts.py ?

**Review by VibhuJawa — CHANGES_REQUESTED** [review]
Thanks for working on this @ryantwolf .  This is a very important functionality we are adding, Very excited for it. 

My major concern currently is around not having a way to rate limit the number of requests we are sending, everything else is mostly nits.

**Review by ayushdg — APPROVED** [review]
Minor nit but at a high level looks good to me! Thanks a lot for this effort

File: nemo_curator/synthetic/async_nemotron.py [hunk 3]
```diff
@@ -0,0 +1,1688 @@
...
+                model=model,
+                subtopic_prompt_template=subtopic_prompt_template,
+                yaml_conversion_prompt_template=yaml_conversion_prompt_template,
+                base_model_kwargs=base_model_kwargs,
+                conversion_model_kwargs=conversion_model_kwargs,
+                ignore_conversion_failure=ignore_conversion_failure,
+            )
+            for macro_topic in macro_topics
+        ]
+        raw_topics = await asyncio.gather(*raw_topics)
```
[the remark this answers is not in the document]
Comment (ryantwolf) [reply nemo_curator/synthetic/async_nemotron.py]: Let me know if you like the way I did it or if you have a better way.

File: nemo_curator/synthetic/nemotron.py (hunk 2 above)
Comment (ryantwolf) [reply nemo_curator/synthetic/nemotron.py]: No, I don't think so. This isn't really useful outside of a conversational formatter imo. I changed the name to be a bit less confusing.

File: nemo_curator/synthetic/async_nemotron.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (VibhuJawa) [reply nemo_curator/synthetic/async_nemotron.py]: This works perfectly, thanks for incorporating changes this quickly.

**Review by VibhuJawa — APPROVED** [review]
LGTM, thanks for working on this Ryan.

File: nemo_curator/synthetic/nemotron.py (hunk 2 above)
Comment (VibhuJawa) [reply nemo_curator/synthetic/nemotron.py]: Ahh, cool, thanks for the variable rename. Threw me off.

**Pull request merged**